picard/config.py

```diff
@@ -129,6 +131,70 @@ def value(self, name, option_type, default=None):
         return default
 
 
+class SettingConfigSection(ConfigSection):
+    """Custom subclass to automatically accommodate saving and retrieving values based on user profile settings.
+    """
+    PROFILES_KEY = 'user_profiles'
+    SETTINGS_KEY = 'user_profile_settings'
+
+    @classmethod
+    def init_profile_options(cls):
+        ListOption.add_if_missing("profiles", cls.PROFILES_KEY, [])
+        Option.add_if_missing("profiles", cls.SETTINGS_KEY, {})
+
+    def __init__(self, config, name):
+        super().__init__(config, name)
+        self.__qt_config = config
+        self.__name = name
+        self.__prefix = self.__name + '/'
+        self._memoization = defaultdict(Memovar)
+        self.init_profile_options()
+
+    def _get_active_profile_ids(self):
+        profiles = self.__qt_config.profiles[self.PROFILES_KEY]
+        if profiles is None:
+            return
+        for profile in profiles:
+            if profile['enabled']:
+                yield profile["id"]
+
+    def _get_active_profile_settings(self):
+        for id in self._get_active_profile_ids():
+            yield id, self._get_profile_settings(id)
+
+    def _get_profile_settings(self, id):
+        profile_settings = self.__qt_config.profiles[self.SETTINGS_KEY][id]
+        if profile_settings is None:
+            log.error("Unable to find settings for user profile '%s'", id)
+            return {}
+        return profile_settings
+
+    def __getitem__(self, name):
+        # Don't process settings that are not profile-specific
+        if name in UserProfileGroups.get_all_settings_list():
+            for id, settings in self._get_active_profile_settings():
+                if name in settings and settings[name] is not None:
+                    return settings[name]
+        opt = Option.get(self.__name, name)
+        if opt is None:
+            return None
+        return self.value(name, opt, opt.default)
+
+    def __setitem__(self, name, value):
+        # Don't process settings that are not profile-specific
+        if name in UserProfileGroups.get_all_settings_list():
+            for id, settings in self._get_active_profile_settings():
+                if name in settings:
+                    settings[name] = value
+                    all_settings = self.__qt_config.profiles[self.SETTINGS_KEY]
+                    all_settings[id] = settings
+                    self.__qt_config.profiles[self.SETTINGS_KEY] = all_settings
```

Comment:
@rdswift I wonder about those 3 lines:

https://github.com/metabrainz/picard/blob/e01b97844132e01c7f5463eef91653cc064c369e/picard/config.py#L189-L191

Aren't they equivalent to:

```python
self.__qt_config.profiles[self.SETTINGS_KEY][id] = settings
```

(I may miss something here)

Reply:
You're not missing something, I was.  I didn't realize that you could update a subscripted value directly.  I'll make the change.